src/primitive_debug.cc

```diff
@@ -60,10 +54,44 @@ PRIMITIVE(object_histogram) {
     } else if (is_string(object) && !String::cast(object)->content_on_heap()) {
       size += String::cast(object)->length() + 1;
     }
-    capture.data[class_index * UINT32_PER_ENTRY + 0] += 1;
-    capture.data[class_index * UINT32_PER_ENTRY + 1] += size;
+    data[class_index].count++;
+    data[class_index].size += size;
   });
-  return result;
+  int non_trivial_entries = 0;
+  int bytes_needed = 0;
+  bool size_known = false;
+  // Encode twice.  First time with a dry run to calculate the size of the
+  // encoded data.
+  for (int attempt = 0; attempt < 2; attempt++) {
+    MallocedBuffer encoding_buffer(size_known ? bytes_needed : 1);
+    if (!encoding_buffer.has_content()) MALLOC_FAILED;
+    ProgramOrientedEncoder encoder(program, &encoding_buffer);
```

Comment:
I think I'd move this out in a helper and call that twice instead of wrapping it in a loop with special handling for both first and second iteration.